File: libgomp/testsuite/libgomp.oacc-fortran/lib-15.f90
! { dg-do run }
! { dg-skip-if "" { *-*-* } { "*" } { "-DACC_MEM_SHARED=0" } }

program main
  use openacc
  implicit none

  integer, parameter :: N = 256
  integer, allocatable :: h(:)
  integer :: i

  allocate (h(N))

  do i = 1, N
    h(i) = i
  end do 

  call acc_copyin (h)

  do i = 1, N
    h(i) = i + i
  end do 

  call acc_update_device (h, sizeof (h))

  if (acc_is_present (h) .neqv. .TRUE.) stop 1

  h(:) = 0

  call acc_copyout (h, sizeof (h))

  do i = 1, N
    if (h(i) /= i + i) stop 2
  end do 

  call acc_copyin (h, sizeof (h))

  h(:) = 0

  call acc_update_self (h, sizeof (h))
  
  if (acc_is_present (h) .neqv. .TRUE.) stop 3

  do i = 1, N
    if (h(i) /= i + i) stop 4
  end do 

  call acc_delete (h)

  if (acc_is_present (h) .neqv. .FALSE.) stop 5
  
end program

File: libgomp/testsuite/libgomp.oacc-c-c++-common/lib-15.c
/* Check acc_is_present and acc_copyout.  */

/* { dg-skip-if "" { *-*-* } { "*" } { "-DACC_MEM_SHARED=0" } } */

#include <stdlib.h>
#include <openacc.h>

int
main (int argc, char **argv)
{
  const int N = 256;
  int i;
  unsigned char *h;

  h = (unsigned char *) malloc (N);

  for (i = 0; i < N; i++)
    {
      h[i] = i;
    }

  (void) acc_copyin (h, N);

  acc_copyout (h, N);

  for (i = 0; i < N; i++)
    {
      if (acc_is_present (h + i, 1) != 0)
	abort ();
    }

  free (h);

  return 0;
}
